File: Binary-Search/1011-capacity-to-ship-packages-within-d-days.py
# Time  : O(nlogm) - Where n is len(weights), m is the (sum - max) in weights
# Space : O(1)

class Solution:
  def shipWithinDays(self, weights: List[int], days: int) -> int:
    lo, hi = max(weights), sum(weights)

    while lo < hi:
      cap = lo + (hi - lo) // 2

      # number of days to ship
      d = 1
      cnt = 0
      for i in range(len(weights)):
        if cnt + weights[i] <= cap:
          cnt += weights[i]
        else:
          d += 1
          cnt = weights[i]

      if d > days:
        lo = cap + 1
      else:
        hi = cap

    return lo

File: Binary-Search/1011-capacity-to-ship-packages-within-d-days.cpp
// Time  : O(nlogm)
// Space : O(1)
class Solution {
public:
  int shipWithinDays(vector<int>& weights, int days) {
    int mx = 0;
    int sum = 0;
    for (int& w : weights) {
      mx = max(mx, w);
      sum += w;
    }
    int lo = mx;
    int hi = sum;
    while (lo < hi) {
      int cap = (lo + hi)>>1;
      int d = 1;
      long long load = 0;
      for (int& w : weights) {
        if (load + w <= cap) load += w;
        else {
          d += 1;
          load = w;
        }
      }
      if (d > days) lo = cap+1;
      else hi = cap;
    }
    return lo;
  }
};